File: member-account-stack-sets/03-password-policy/password-policy.js
'use strict';
const AWS = require('aws-sdk');
const response = require('cfn-response');
const iam = new AWS.IAM({apiVersion: '2010-05-08'});
exports.handler = (event, context, cb) => {
    console.log(`Invoke: ${JSON.stringify(event)}`);
    const done = (err) => {
        if (err) {
            console.log(`Error: ${JSON.stringify(err)}`);
            response.send(event, context, response.FAILED, {});
        } else {
            response.send(event, context, response.SUCCESS, {});
        }
    };
    if (event.RequestType === 'Delete') {
        iam.deleteAccountPasswordPolicy({}, done);
    } else if (event.RequestType === 'Create' || event.RequestType === 'Update') {
        iam.updateAccountPasswordPolicy({
            AllowUsersToChangePassword: Boolean(event.ResourceProperties.AllowUsersToChangePassword),
            HardExpiry: Boolean(event.ResourceProperties.HardExpiry),
            MaxPasswordAge: event.ResourceProperties.MaxPasswordAge,
            MinimumPasswordLength: event.ResourceProperties.MinimumPasswordLength,
            PasswordReusePrevention: event.ResourceProperties.PasswordReusePrevention,
            RequireLowercaseCharacters: Boolean(event.ResourceProperties.RequireLowercaseCharacters),
            RequireNumbers: Boolean(event.ResourceProperties.RequireNumbers),
            RequireSymbols: Boolean(event.ResourceProperties.RequireSymbols),
            RequireUppercaseCharacters: Boolean(event.ResourceProperties.RequireUppercaseCharacters),
        }, done);
    } else {
        cb(new Error(`unsupported RequestType: ${event.RequestType}`));
    }
};

File: member-account-stack-sets/03-password-policy/password-policy.py
import cfnresponse
import boto3

iam = boto3.client('iam')

def is_bool(value):
    return value.lower() == 'true'

password_policy_keys = dict(
    MinimumPasswordLength=int,
    RequireSymbols=is_bool,
    RequireNumbers=is_bool,
    RequireUppercaseCharacters=is_bool,
    RequireLowercaseCharacters=is_bool,
    AllowUsersToChangePassword=is_bool,
    MaxPasswordAge=int,
    PasswordReusePrevention=int,
    HardExpiry=is_bool
)

def handler(event, context):
    print(event)
    try:
        resource_properties = event['ResourceProperties']
        request_type = event['RequestType']
        if request_type in ['Create', 'Update']:
            update_parameters = {key: cast_type(resource_properties[key]) for key, cast_type in
                                 password_policy_keys.items()}
            print(update_parameters)
            response = iam.update_account_password_policy(**update_parameters)
            print(response)
        elif request_type is 'Delete':
            iam.delete_account_password_policy()

        cfnresponse.send(event, context, cfnresponse.SUCCESS, {}, "")
    except Exception as e:
        print(e)
        cfnresponse.send(event, context, cfnresponse.FAILED, {}, "")
